doc/monads/except.lean

```diff
@@ -0,0 +1,135 @@
+    throw "can't divide by zero"
+  else
+    pure (x / y)
+
+#eval divide 5 2  -- Except.ok 2.500000
+#eval divide 5 0  -- Except.error "can't divide by zero"
+
+/-!
+
+Just as the `read` operation was available from the `ReaderM` monad and the `get` and `set`
+operations came with the `StateM` monad, here you can see a `throw` operation is provided by the
+`Except` monad.
+
+So in Lean, `throw` is not available everywhere like it is in most imperative programming languages.
+You have to declare your function can throw by changing the type signature to `Except String Float`.
+This creates a function that might return an error of type `String` or it might return a value of
+type `Float` in the non-error case.
+
+Once your function is monadic you also need to use the `pure` constructor of the `Except` monad to
+convert the pure non-monadic value `x / y` into the required `Except` object.  See
+[Applicatives](applicatives.lean.md) for details on `pure`.
+
+Now this return typing would get tedious if you had to include it everywhere that you call this
+function, however, Lean type inference can clean this up. For example, you can define a test
+function can calls the `divide` function and you don't need to say anything here about the fact that
+it might throw an error, because that is inferred:
+-/
+def test := divide 5 0
+
+#check test     -- Except String Float
+/-!
+
+Notice the Lean compiler infers the required `Except String Float` type information for you.
+And now you can run this test and get the expected exception:
+
+-/
+#eval test      -- Except.error "can't divide by zero"
+/-!
+
+## Chaining
+
+Now as before you can build a chain of monadic actions that can be composed together using `bind (>>=)`:
+-/
+def square (x : Float) : Except String Float :=
+  if x >= 100 then
+    throw "it's absolutely huge"
+  else
+    pure (x * x)
+
+#eval divide 6 2 >>= square  -- Except.ok 9.000000
+#eval divide 6 0 >>= square  -- Except.error "can't divide by zero"
+#eval divide 100 1 >>= square  -- Except.error "it's absolutely huge"
+
+def chainUsingDoNotation := do
+  let r ← divide 6 0
+  square r
+
+#eval chainUsingDoNotation  -- Except.error "can't divide by zero"
+
+/-!
+Notice in the second `divide 6 0` the exception from that division was nicely propagated along
+to the final result and the square function was pretty much ignored in that case.
+
+Remember also that you can chain the actions with implicit binding by using the `do` notation
+as you see in the `chainUsingDoNotation` function above.
+
+## Try/Catch
+
+Now with all good exception handling you also want to be able to catch exceptions so your program
+can try continue on or do some error recovery task, which you can do like this:
+-/
+def testCatch :=
+  try
+    let r ← divide 8 0  -- 'r' is type Float
+    pure (toString r)
+  catch e =>
+    pure s!"Caught exception: {e}"
+/-!
+
+Note that the type inferred by Lean for this function is `Except String String` so the `Except
```

Comment:
I am not sure what you are trying to say here. What transformation are you referring to? 
There is no magic happening here, you have used `toString` to convert the `Float` into a `String`.

Reply:
Yeah it is subtle and I ran into a ton of confusion in this area which is why I'm calling this out.  If I wrote this:

```lean
def test := divide 5 0
```
lean would have inferred the type for `test` to be `Except String Float`.  But if I change the pure return type to a string as in testCatch, then lean magically figures out that I must have wanted the type for `testCatch` to be `Except String String`, without me having to state that explicitly.  

We had a zulip discussion about this, or was it in the last wed call, where I was wondering why `testCatch` didn't infer the return type of `String` since I caught the exception and explicitly made the catch block convert the error to a string instead of rethrowing.  You said lean chose to continue to propagate the Except type in this case because someone "might" choose to rethrow in a case like this... anyway this is what I'm trying to explain.  Any thoughts on how to improve it?

Reply:
but perhaps "transformation" is the wrong language, perhaps I should just say it is another kind of "type inference".

Reply:
> But if I change the pure return type to a string as in testCatch, then lean magically figures out that I must have wanted the type for testCatch to be Except String String

Let's focus on the `try` block you wrote in this example. It is
```lean
    let r ← divide 8 0  -- 'r' is type Float
    pure (toString r)
```
Which is notation for
```lean
bind (divide 8 0) (fun (r : Float) => pure (toString r))
```
We have 
```lean
divide 8 0 : Except String Float
```
and 
```lean
fun (r : Float) => pure (toString r) : Fload -> Except String String
```
Recall that `toString r` has type `String`, and `pure (toString r) : Except String String`.
Now, the `bind` puts together these two "pieces" together that have types `Except String Float` and `Float -> Except String String`, and produces a value of type `Except String String`. 
There is no magic.

Reply:
> but perhaps "transformation" is the wrong language, perhaps I should just say it is another kind of compositional "type inference".

There is no such thing in Lean.

Reply:
Fixed in https://github.com/leanprover/lean4/pull/1505/commits/9e2398bf897e5bd656ec1ad48b83d32896079e6a

Reply:
Why leave this to inference anyway? We always give explicit types of top-level declarations, is there any reason not to do that here?